### PR TITLE
CPDP-793: amend email

### PR DESCRIPTION
**Pull request by co-category**
@AndyHWChung please could you review these early draft PRs. If you agree with the approach then I will proceed with the tests.

### CODE REVIEW

File: app/uk/gov/hmrc/cgtpropertydisposalsfrontend/controllers/email/RegistrationChangeEmailController.scala [hunk 1]
```diff
@@ -30,10 +31,11 @@ import uk.gov.hmrc.cgtpropertydisposalsfrontend.models.name.ContactName
 import uk.gov.hmrc.cgtpropertydisposalsfrontend.repos.SessionStore
 import uk.gov.hmrc.cgtpropertydisposalsfrontend.services.EmailVerificationService
 import uk.gov.hmrc.cgtpropertydisposalsfrontend.util.Logging
-import uk.gov.hmrc.cgtpropertydisposalsfrontend.{controllers, views}
+import uk.gov.hmrc.cgtpropertydisposalsfrontend.{controllers, models, views}
+import uk.gov.hmrc.http.HeaderCarrier
 import uk.gov.hmrc.play.bootstrap.controller.FrontendController
-
-import scala.concurrent.ExecutionContext
+import cats.implicits._
```
Comment (AndyHWChung) [inline app/uk/gov/hmrc/cgtpropertydisposalsfrontend/controllers/email/RegistrationChangeEmailController.scala]: I think you can scope this down to
```
import cats.instances.future._
```

File: app/uk/gov/hmrc/cgtpropertydisposalsfrontend/controllers/email/SubscriptionChangeEmailController.scala [hunk 2]
```diff
@@ -53,29 +56,30 @@ class SubscriptionChangeEmailController @Inject()(
...
+  ): Either[Result, (SessionData, Subscribed)] =
     validJourney(request)
 
-  val subscriptionReadyEmailLens: Lens[SubscriptionReady, Email] =
-    lens[SubscriptionReady].subscriptionDetails.emailAddress
+  override def updateEmail(journey: Subscribed, email: Email)(
+    implicit hc: HeaderCarrier
+  ): EitherT[Future, Error, Subscribed] =
+    subscriptionService
+      .updateSubscribedDetails(journey.subscribedDetails)
```
Comment (AndyHWChung) [inline app/uk/gov/hmrc/cgtpropertydisposalsfrontend/controllers/email/SubscriptionChangeEmailController.scala]: wouldn't you need to do something like `journey.subscribedDetails.copy(email = email)` here?

File: app/uk/gov/hmrc/cgtpropertydisposalsfrontend/controllers/email/RegistrationEnterEmailController.scala [hunk 3]
```diff
@@ -31,9 +32,10 @@ import uk.gov.hmrc.cgtpropertydisposalsfrontend.repos.SessionStore
 import uk.gov.hmrc.cgtpropertydisposalsfrontend.services.EmailVerificationService
 import uk.gov.hmrc.cgtpropertydisposalsfrontend.util.Logging
 import uk.gov.hmrc.cgtpropertydisposalsfrontend.{controllers, views}
+import uk.gov.hmrc.http.HeaderCarrier
 import uk.gov.hmrc.play.bootstrap.controller.FrontendController
-
-import scala.concurrent.ExecutionContext
+import cats.implicits._
```
Comment (AndyHWChung) [inline app/uk/gov/hmrc/cgtpropertydisposalsfrontend/controllers/email/RegistrationEnterEmailController.scala]: again, I think you should be able to just import `cats.instances.future._`

File: app/uk/gov/hmrc/cgtpropertydisposalsfrontend/controllers/email/SubscriptionChangeEmailController.scala (hunk 2 above)
Comment (AndyHWChung) [reply app/uk/gov/hmrc/cgtpropertydisposalsfrontend/controllers/email/SubscriptionChangeEmailController.scala]: but also, this is the wrong controller to do this in. We need a new controller for changing emails when subscribed - this one if for the subscription journey. We need something like `SubscribedChangeEmailController` which has the backlinks and continueUrl pointing to the homepage. Then we need routes for those endpionts in `app.routes` and the homepage template then needs the change email link stuck in there

Comment (co-category) [reply app/uk/gov/hmrc/cgtpropertydisposalsfrontend/controllers/email/SubscriptionChangeEmailController.scala]: OK, thanks for the feedback. When we discussed this originally we were looking at the SubscriptionChangeEmailController file and I thought I had to change that. But this makes sense I will change it.

File: app/uk/gov/hmrc/cgtpropertydisposalsfrontend/controllers/email/EmailController.scala [hunk 4]
```diff
@@ -158,21 +161,20 @@ trait EmailController[Journey <: JourneyStatus, VerificationCompleteJourney <: J
...
-                      logger.warn("Could not store email verified result", e)
-                      errorHandler.errorResult()
-                    }, { _ =>
-                      Redirect(emailVerifiedCall)
-                    }
-                  )
+                val email = Email(emailToBeVerified.email.value)
+                val result: EitherT[Future, Result, Unit] = for {
+                  journey <- updateEmail(journey, email).leftMap(e => errorHandler.errorResult())
+                  _ <- EitherT.liftF(updateSession(sessionStore, request) { s =>
```
[the remark this answers is not in the document]
Comment (co-category) [reply app/uk/gov/hmrc/cgtpropertydisposalsfrontend/controllers/email/EmailController.scala]: fixed.

File: app/uk/gov/hmrc/cgtpropertydisposalsfrontend/controllers/email/RegistrationChangeEmailController.scala (hunk 1 above)
Comment (co-category) [reply app/uk/gov/hmrc/cgtpropertydisposalsfrontend/controllers/email/RegistrationChangeEmailController.scala]: fixed

File: app/uk/gov/hmrc/cgtpropertydisposalsfrontend/controllers/email/SubscriptionChangeEmailController.scala (hunk 2 above)
Comment (co-category) [reply app/uk/gov/hmrc/cgtpropertydisposalsfrontend/controllers/email/SubscriptionChangeEmailController.scala]: created the new controller class with the back link and continue link set appropriately

File: app/uk/gov/hmrc/cgtpropertydisposalsfrontend/controllers/email/RegistrationEnterEmailController.scala (hunk 3 above)
Comment (co-category) [reply app/uk/gov/hmrc/cgtpropertydisposalsfrontend/controllers/email/RegistrationEnterEmailController.scala]: fixed

File: app/uk/gov/hmrc/cgtpropertydisposalsfrontend/controllers/email/SubscribedChangeEmailController.scala [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    validJourney(request)
+
+  override def updateEmail(journey: Subscribed, email: Email)(
+    implicit hc: HeaderCarrier
+  ): EitherT[Future, Error, Subscribed] = {
+    val journeyWithUpdatedEmail = journey.subscribedDetails.copy(emailAddress = email)
+    subscriptionService
+      .updateSubscribedDetails(journeyWithUpdatedEmail)
+      .map(_ => journey.copy(journeyWithUpdatedEmail))
+  }
```
Comment (co-category) [inline app/uk/gov/hmrc/cgtpropertydisposalsfrontend/controllers/email/SubscribedChangeEmailController.scala]: Is there a better way to do this?

**Comment by co-category** [comment]
@AndyHWChung pls review again (tests pending on review)

File: app/uk/gov/hmrc/cgtpropertydisposalsfrontend/views/account/manage_your_details.scala.html [hunk 6]
```diff
@@ -76,7 +76,7 @@ <h2 class="heading-medium">
   @cyaRow(
     messages("subscription.cyaLabel.email"),
     Html(details.emailAddress.value),
-    Some(cyaChange(messages("subscription.cyaChange.email"), "#"))
+    Some(cyaChange(messages("subscription.cyaChange.email"), "/cgt-property-disposals/subscribed/amend-details/change-email")))
```
Comment (AndyHWChung) [inline app/uk/gov/hmrc/cgtpropertydisposalsfrontend/views/account/manage_your_details.scala.html]: you can use the reverse routing stuff here to be able to cope better with possible future changes to the url, so you could have `email.routes.BlahController.blah` here

File: app/uk/gov/hmrc/cgtpropertydisposalsfrontend/controllers/email/EmailController.scala [hunk 7]
```diff
@@ -158,21 +161,20 @@ trait EmailController[Journey <: JourneyStatus, VerificationCompleteJourney <: J
...
-                }.map(
-                  _.fold(
-                    { e =>
-                      logger.warn("Could not store email verified result", e)
-                      errorHandler.errorResult()
-                    }, { _ =>
-                      Redirect(emailVerifiedCall)
-                    }
-                  )
+                val email = Email(emailToBeVerified.email.value)
```
[the remark this answers is not in the document]
Comment (co-category) [reply app/uk/gov/hmrc/cgtpropertydisposalsfrontend/controllers/email/EmailController.scala]: Nope, it's 

`final case class EmailToBeVerified(email: Email, id: UUID, verified: Boolean)`

[the remark this answers is not in the document]
Comment (AndyHWChung) [reply app/uk/gov/hmrc/cgtpropertydisposalsfrontend/controllers/email/EmailController.scala]: oh. You should be able to just use `emailToBeVerified.email` here then right?

File: test/uk/gov/hmrc/cgtpropertydisposalsfrontend/controllers/email/EmailControllerSpec.scala [hunk 8]
```diff
@@ -50,7 +50,9 @@ trait EmailControllerSpec[Journey <: JourneyStatus, VerificationCompleteJourney
 
   val validVerificationCompleteJourneyStatus: VerificationCompleteJourney
 
-  def updateEmail(journey: Journey, email: Email): VerificationCompleteJourney
+  def updateEmail(journey: Journey, email: Email)(
+    implicit hc: HeaderCarrier
+  ): EitherT[Future, Error, VerificationCompleteJourney]
```
Comment (AndyHWChung) [inline test/uk/gov/hmrc/cgtpropertydisposalsfrontend/controllers/email/EmailControllerSpec.scala]: with this type signature you lose the ability to test what happens if the update of the email fails. One possible way to be able to control that in the tests is to leave `updateEmail` as it was and add a new `val` which looks something like:
```scala
val mockUpdateEmail: Option[(VerificationJourney,EitherT[Future,Error,Unit]) => Unit]
```
The `Option` allows you to specifiy in the implementing tests whether or not an email update can fail. You can add an extra test which tests failure on email updates now as well:
```scala
  "return an error" when {
    mockUpdateEmail.foreach{ f =>
      "the email update fails" in {
          ...
       }
   }
  }
```

Comment (co-category) [reply test/uk/gov/hmrc/cgtpropertydisposalsfrontend/controllers/email/EmailControllerSpec.scala]: Fixed tests as discussed.

**Comment by AndyHWChung** [comment]
LGTM

**Pull request merged**